### DISCUSSION TITLE
Updating Routes and Cluster without Restarting Yarp

### DISCUSSION THREAD

**Discussion by Octania** [dotnet/yarp#2310] [Q&A] [answered]
I read this some place but can't remember.
Yarp/service does not require a restarts when the configuration in appSettings.json changes.
For example: I am running Yarp and changed a route then I do not have to restart, that route simply becomes  active.

My questions is, what about an XML config outside of the conventional appSettings.json.

I can watch the file for changes but wondering how I can make YARP bind to any route changes if an XML config is changed?
Is this possible?

There is this documentation but not very clear:
https://microsoft.github.io/reverse-proxy/articles/config-providers.html#reload

Any examples, or reference is highly appreciated.

Thanks

**Comment by Tratcher** [comment]
How are you binding to the xml the first time?

**Reply by Tratcher** [reply]
See this example as one way to trigger updates:
https://github.com/microsoft/reverse-proxy/blob/3a5a68fac3bccbcf19e785a82497abf4b1c30053/samples/ReverseProxy.Code.Sample/Startup.cs#L31
https://github.com/microsoft/reverse-proxy/blob/3a5a68fac3bccbcf19e785a82497abf4b1c30053/samples/ReverseProxy.Code.Sample/Startup.cs#L44

**Comment by Octania** [comment]
> How are you binding to the xml the first time?

@Tratcher 
I am implementing the ` IProxyConfigProvider` interface and `GetConfig()`.
Then in the implementing class I am creating or binding the new routes by simply reading their values from the file.

**Like this:**

```
public class ProxyConfigProvider : IProxyConfigProvider
    {
        private readonly ProxyConfig _config;
        public ProxyConfigProvider()
        {
            _config = new ProxyConfig(ConfigureRoute(), ConfigureCluster());
        }

        public IProxyConfig GetConfig() => _config;
        private IReadOnlyList<RouteConfig> ConfigureRoute()
        {
            return new[]
             {
                 new RouteConfig
                 {
                     // New Route from XML
                 }
             };
        }

        private static IReadOnlyList<ClusterConfig> ConfigureCluster()
        {
            return new[]
                {
                    new ClusterConfig
                    {
                        Destinations = new Dictionary<string, DestinationConfig>()
                        {
                            // New Distination here
                        },
                    },
                };
        }
    }
```

**Call it like this in program.cs:**
`            services.AddSingleton<IProxyConfigProvider>(new ProxyConfigProvider()).AddReverseProxy();`

**Reply by Tratcher** [reply, marked as answer]
Ok. Take a look at how we read the appsettings.json and listen for changes:
https://github.com/microsoft/reverse-proxy/blob/3a5a68fac3bccbcf19e785a82497abf4b1c30053/src/ReverseProxy/Configuration/ConfigProvider/ConfigurationConfigProvider.cs#L52-L62

You should be able to do something similar with FileSystemWatcher.
https://learn.microsoft.com/en-us/dotnet/api/system.io.filesystemwatcher?view=net-7.0

**Reply by Octania** [reply]
Cool will look into this.
Thanks for responding,
